File: sem3/matr/b.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
    freopen("destroy.in", "r", stdin);
    freopen("destroy.out", "w", stdout);

    long long n, m, s;
    cin >> n >> m >> s;
    vector<vector<tuple<long long, long long, long long>>> g(n + 1);
    map<long long, long long> d;
    for (long long i = 1; i <= m; ++i) {
        long long a, b, c;
        cin >> a >> b >> c;
        g[a].push_back({c, b, i});
        g[b].push_back({c, a, i});
        d[i] = c;
    }
    set<long long> v;
    v.insert(n);
    priority_queue<tuple<long long, long long, long long>> q;
    for (auto & i : g[n]) {
        q.push(i);
    }
    tuple<long long, long long, long long> t = {0, 0, 0};
    while (v.size() < n) {
        while (!q.empty()) {
            t = q.top();
            q.pop();
            if (v.count(get<1>(t)) == 0) {
                break;
            }
        }
        v.insert(get<1>(t));
        for (auto & i : g[get<1>(t)]) {
            q.push(i);
        }
        d.erase(get<2>(t));
    }
    set<pair<long long, long long>> p;
    for (auto & i : d) {
        p.insert({i.second, i.first});
    }
    vector<long long> r;
    for (auto e = p.begin(); e != p.end() && s >= e->first; ++e) {
        s -= e->first;
        r.push_back(e->second);
    }
    sort(r.begin(), r.end());
    cout << r.size() << '\n';
    for (long long i = 0; i < r.size(); ++i) {
        cout << r[i] << ' ';
    }
}

File: sem3/matr/b.py
import sys, queue


sys.stdin = open('destroy.in', 'r')
sys.stdout = open('destroy.out', 'w')

n, m, s = map(int, input().split())
g = [[] for _ in range(n + 1)]
d = dict()
for i in range(1, m + 1):
    a, b, c = map(int, input().split())
    g[a].append([-c, b, i])
    g[b].append([-c, a, i])
    d[i] = c
v = set()
v.add(1)
q = queue.PriorityQueue()
for i in g[1]:
    q.put(i)
t = [0, 0, 0]
while len(v) < n:
    while not q.empty():
        t = q.get()
        if not (t[1] in v):
            break
    v.add(t[1])
    for i in g[t[1]]:
        q.put(i)
    d.pop(t[2])
r = []
for e in sorted(d.items(), key=lambda x: x[1]):
    if s < e[1]:
        break
    s -= e[1]
    r.append(e[0])
r.sort()
print(len(r))
print(*r)
